control-plane/grpc/src/operations/jsongrpc/client.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{
+    context::{Client, Context, TracedChannel},
+    jsongrpc::{json_grpc_client, json_grpc_reply, ProbeRequest},
+    operations::jsongrpc::traits::{JsonGrpcOperations, JsonGrpcRequestInfo},
+};
+use common_lib::{
+    mbus_api::{ReplyError, ResourceKind, TimeoutOptions},
+    types::v0::message_bus::MessageIdVs,
+};
+use serde_json::Value;
+use std::{ops::Deref, time::Duration};
+use tonic::transport::Uri;
+
+/// RPC JsonGrpc Client
+#[derive(Clone)]
+pub struct JsonGrpcClient {
+    inner: Client<json_grpc_client::JsonGrpcClient<TracedChannel>>,
+}
+
+impl JsonGrpcClient {
+    /// creates a new base tonic endpoint with the timeout options and the address
+    pub async fn new<O: Into<Option<TimeoutOptions>>>(addr: Uri, opts: O) -> Self {
+        let client = Client::new(addr, opts, json_grpc_client::JsonGrpcClient::new).await;
+        Self { inner: client }
+    }
+    /// Try to wait until the JsonGrpc Service is ready, up to a timeout, by using the Probe method.
+    pub async fn wait_ready(&self, timeout_opts: Option<TimeoutOptions>) -> Result<(), ()> {
+        let timeout_opts = match timeout_opts {
+            Some(opts) => opts,
+            None => TimeoutOptions::new()
+                .with_timeout(Duration::from_millis(250))
+                .with_max_retries(10),
+        };
+        for attempt in 1 ..= timeout_opts.max_retries().unwrap_or_default() {
+            match self
+                .probe(Some(Context::new(Some(timeout_opts.clone()))))
+                .await
+            {
+                Ok(true) => return Ok(()),
+                _ => {
+                    let delay = std::time::Duration::from_millis(100);
+                    tracing::trace!(%attempt, delay=?delay, "Not available, retrying after...");
```

Comment:
this trace message is incorrect in the last iteration, as it won't be tried after..

Reply:
Right. Would skip it in the last iteration.